source/server/listener_impl.cc

```diff
@@ -106,6 +129,9 @@ ListenerImpl::ListenerImpl(const envoy::api::v2::Listener& config, const std::st
       listener_scope_(
           parent_.server_.stats().createScope(fmt::format("listener.{}.", address_->asString()))),
       bind_to_port_(PROTOBUF_GET_WRAPPED_OR_DEFAULT(config.deprecated_v1(), bind_to_port, true)),
+      socket_type_(Network::Utility::protobufAddressSocketType(config.address())),
+      reuse_port_((socket_type_ == Network::Address::SocketType::Datagram) ? true
```

Comment:
I highly suspect that we don't need to latch these two parameters. socket_type_ is only used in this constructor and then it is latched in socket factory already. Though reuse_port_ is used in createListenSocketFactory() via ListenerImpl::reusePort(), I think the caller of this method actually has direct access to listener protobuf. Can you try to remove these two members and reusePort() from this impl?

Reply:
I didn't mean that you always get them from protobuf... These two variables is only needed in this constructor or the caller of it. So they can just be local variable.

Reply:
reasons why they are members:
- ListenerImpl::reusePort() returns `reuse_port_`.
-  `reuse_port_` is defined as const 
- `reuse_port_` must be inited in ctor, it depends on socket_type, so socket_type is lifted as a member
- ListenerManagerImpl::createListenSocketFactory() calls  ListenerImpl::reusePort()  to create ListenSocketFactoryImpl

PS: Just thought if a class should have a member not only determined by in how many methods it's used, socket_type and reuse_port are attributes of a listener, they could be used somewhere later.

Reply:
It's a bit confusing to latch same value in two different places. reuse_port_ and socket_type_ is already stored in ListenSocketFactory. And later we might need to maintain them in sync. That's why I want to remove these two members from ListenerImpl. If we need to use them somewhere else, we can access them by listener->listenSocketFactory()->reusePort().

Originally socket_type was a local var. 

The only need of reuse_port here is to add SO_REUSEPORT option. You can just embed `(socket_type_ == Network::Address::SocketType::Datagram) ? true : config.reuse_port()` in the if condition.
You can also change ListenerManagerImpl::createListenSocketFactory() to take reuse_port as parameter, and its caller  ListenerManagerImpl::addOrUpdateListener() should also have access to config. So ListenerImpl doesn't need reusePort() interface.

Reply:
+1 sounds reasonable to me.

/wait

Reply:
I don't think this causes confusion and need additional effort to sync them. If it really causes any issue, such kind of issue is easy to fix later, let finish this PR, ok? ;)

Reply:
@l8huang I agree with @danzh2010 on this. Please make the change. Thanks!

/wait

Reply:
@mattklein123 @danzh2010 please take a look, changed according to comment.